### PR TITLE
fix for dragged table name

### LINKED ISSUE #11372: Object Explorer drag-and-drop table incorrectly wraps table names
- Azure Data Studio Version: `1.20.0

Steps to Reproduce:

1. Drag a table from Object Explorer to the query editor 

Expected: Tables are wrap with brackets around each part, for example `[database].[schema].[table]`.

Actual: Table are wrapped incorrectly, for example , `[schema].[table]`

Screen Shot 2020-07-15 at 7 51 50 PM

### PR DESCRIPTION
**Pull request by cssuh**
Fixed the formatting of the dragged table name

This PR fixes #11372

### CODE REVIEW

File: src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts [hunk 1]
```diff
@@ -76,7 +76,7 @@ export class ServerTreeDragAndDrop implements IDragAndDrop {
 		if (element.nodeTypeId === 'Column' || element.nodeTypeId === 'Table') {
 			const schema = element.metadata.schema;
 			const name = element.metadata.name;
-			originalEvent.dataTransfer.setData(DataTransfers.RESOURCES, JSON.stringify([`${element.nodeTypeId}:${element.id}?${schema ? schema + '.' + name : name}`]));
+			originalEvent.dataTransfer.setData(DataTransfers.RESOURCES, JSON.stringify([`${element.nodeTypeId}:${element.id}?${schema ? schema + '].[' + name : name}`]));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts:79]: You need to escape ]'s in the names

**Comment by cssuh** [comment]
ah okay I'll do that

**Comment by Charles-Gagnon** [comment]
No - the [ being added starts the quoted string and so anything up to the next closing bracket will be considered quoted. You can't have nested quoted strings - any other ['s inside will be treated as part of that string. This behaves just like quoted strings in most other languages - think of how you have to escape " for normal strings for example.

Make sure when you're making your changes that you're testing this yourself as well as users would - so try inserting it into statements and executing them. And adding some tests around these specific cases isn't a bad idea either.

File: src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts [hunk 2]
```diff
@@ -74,9 +74,9 @@ export class ServerTreeDragAndDrop implements IDragAndDrop {
 		const data = dragAndDropData.getData();
 		const element = data[0];
 		if (element.nodeTypeId === 'Column' || element.nodeTypeId === 'Table') {
-			const schema = element.metadata.schema;
-			const name = element.metadata.name;
-			originalEvent.dataTransfer.setData(DataTransfers.RESOURCES, JSON.stringify([`${element.nodeTypeId}:${element.id}?${schema ? schema + '.' + name : name}`]));
+			const escapedSchema = element.metadata.schema?.replace(/]/g, ']]');
+			const escapedName = element.metadata.name?.replace(/]/g, ']]');
+			originalEvent.dataTransfer.setData(DataTransfers.RESOURCES, JSON.stringify([`${element.nodeTypeId}:${element.id}?${escapedSchema ? escapedSchema + '].[' + escapedName : escapedName}`]));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts:79]: And if we're going to be adding the middle brackets here then I don't think we should be doing the outer brackets here : https://github.com/microsoft/azuredatastudio/pull/10702/files#diff-d38769f349f31d56f614f8337e3a8965R363

Is there any reason you couldn't just do the whole quoted name here?

Comment (aasimkhan30) [reply src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts:79]: Shouldn't we have this escape logic in some common place so other parts of ads can access them?

File: src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts [hunk 3]
```diff
@@ -74,9 +74,10 @@ export class ServerTreeDragAndDrop implements IDragAndDrop {
 		const data = dragAndDropData.getData();
 		const element = data[0];
 		if (element.nodeTypeId === 'Column' || element.nodeTypeId === 'Table') {
-			const schema = element.metadata.schema;
-			const name = element.metadata.name;
-			originalEvent.dataTransfer.setData(DataTransfers.RESOURCES, JSON.stringify([`${element.nodeTypeId}:${element.id}?${schema ? schema + '.' + name : name}`]));
+			const escapedSchema = element.metadata.schema?.replace(/]/g, ']]');
+			const escapedName = element.metadata.name?.replace(/]/g, ']]');
+			const finalString = escapedSchema ? '[' + escapedSchema + '].[' + escapedName + ']' : '[' + escapedName + ']';
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts:79]: Nit - easier to read if you just use interpolated strings

```
const finalString = escapedSchema ? `[${escapedSchema}].[${escapedName}]` : `[${escapedName}]`;
```

**Comment by alanrenmsft** [comment]
@cssuh please merge this PR and create a port request for release/1.20 branch

File: src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/services/objectExplorer/browser/dragAndDropController.ts]: Potentially useful. Not really a huge deal but yeah having something that escapes and/or quotes identifiers might be useful enough to warrant making a helper method for.

**Pull request merged**